site/content/blog/incorporating-external-resources-in-kapp.md

```diff
@@ -0,0 +1,161 @@
+ns-sample  configmap-sample  ConfigMap  -       -    create  -       reconcile  -   -
+^          secret-sample     Secret     -       -    create  -       reconcile  -   -
+
+Op:      2 create, 0 delete, 0 update, 0 noop
+Wait to: 2 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: y
+
+9:24:14AM: ---- applying 2 changes [0/2 done] ----
+9:24:14AM: create secret/secret-sample (v1) namespace: ns-sample
+
+kapp: Error: Applying create secret/secret-sample (v1) namespace: ns-sample:
+  Creating resource secret/secret-sample (v1) namespace: ns-sample: namespaces "ns-sample" not found (reason: NotFound)
+```
+
+We know that the namespace will be created by a controller at some point of time, but how do we make sure that we are creating our resources after the namespace has been created. Should we wait for the creation of the namespace and then deploy our app? Sounds mundane, right? Well, we do have a solution for this which will allow us to deploy the app and let kapp do the waiting on it's own. We will use the `kapp.k14s.io/exists` annotation for this.
+
+## How to use the _exists_ annotation
+To use the _exists_ annotation, we would need the basic information of the resource that we want to wait for. By basic, we mean the information that uniquely identifies the resource. In our case, we would need the _apiVersion_, _kind_ and _name_ of a namespace to wait for it.
+Next step is to add this information in the yaml along with the other resources and add the exists annotation to it. So, our yaml would now look something like this.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: ns-sample
+  annotations:
+    kapp.k14s.io/exists: ""
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: configmap-sample
+  namespace: ns-sample
+data:
+  player_initial_lives: "3"
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: secret-sample
+  namespace: ns-sample
+data:
+  extra: YmFyCg==
+```
+
+If we deploy the application now, we will notice that kapp would start waiting for the namespace to be created. Meanwhile we can open another tab in our terminal and create the namespace, and we would notice that kapp would then proceed on to creating the configmap and the secret.
+
+```bash
+kapp deploy -a test -f config.yml
+Target cluster 'https://192.168.64.53:8443' (nodes: minikube)
+
+Changes
+
+Namespace  Name              Kind       Conds.  Age  Op      Op st.  Wait to    Rs  Ri
+(cluster)  ns-sample         Namespace  -       -    exists  -       reconcile  -   -
+ns-sample  configmap-sample  ConfigMap  -       -    create  -       reconcile  -   -
+^          secret-sample     Secret     -       -    create  -       reconcile  -   -
+
+Op:      2 create, 0 delete, 0 update, 0 noop, 1 exists
+Wait to: 3 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: y
+
+10:11:24AM: ---- applying 1 changes [0/3 done] ----
+10:11:24AM: exists namespace/ns-sample (v1) cluster
+10:11:24AM:  ^ Retryable error: External resource doesnt exists
```

Comment:
nit: doesnt should be doesn't

Reply:
I had to remove the single quote as it was messing up with the bash formatting, I will see how we can use escape sequence for quotes in a code sample in markdown.

Reply:
How about making it `does not exist` 🤔

Reply:
That does sound like a good alternative. Thanks :)